File: java/Hill.java
public class Hill {

    // Finds modular inverse of a number
    private static int modInverse(int a, int m) {
        a = a % m;
        for (int x = 1; x < m; x++) {
            if ((a * x) % m == 1) {
                return x;
            }
        }
        return -1;
    }

    // Calculates the inverse of a 2x2 matrix modulo 26
    private static int[][] getInverseMatrix(int[][] key) {
        int det = (key[0][0] * key[1][1] - key[0][1] * key[1][0]) % 26;
        if (det < 0)
            det += 26;

        int detInv = modInverse(det, 26);
        if (detInv == -1) {
            throw new IllegalArgumentException("Key matrix is not invertible.");
        }

        int[][] invKey = new int[2][2];

        // Adjugate matrix * detInv
        // [ d -b ]
        // [ -c a ]
        invKey[0][0] = (key[1][1] * detInv) % 26;
        invKey[0][1] = ((-key[0][1] % 26 + 26) * detInv) % 26;
        invKey[1][0] = ((-key[1][0] % 26 + 26) * detInv) % 26;
        invKey[1][1] = (key[0][0] * detInv) % 26;

        return invKey;
    }

    // Encrypts or Decrypts based on the matrix provided
    private static String process(String text, int[][] key) {
        StringBuilder result = new StringBuilder();
        text = text.toUpperCase().replaceAll("[^A-Z]", "");

        if (text.length() % 2 != 0)
            text += "X"; // Padding

        for (int i = 0; i < text.length(); i += 2) {
            int p1 = text.charAt(i) - 'A';
            int p2 = text.charAt(i + 1) - 'A';

            // Matrix multiplication
            int c1 = (key[0][0] * p1 + key[0][1] * p2) % 26;
            int c2 = (key[1][0] * p1 + key[1][1] * p2) % 26;

            result.append((char) (c1 + 'A'));
            result.append((char) (c2 + 'A'));
        }
        return result.toString();
    }

    public static void main(String[] args) {
        // Key Matrix:
        // 3 3
        // 2 5
        int[][] key = { { 3, 3 }, { 2, 5 } };
        String message = "HELP";

        // Encryption
        String encrypted = process(message, key);
        System.out.println("Encrypted: " + encrypted);

        // Decryption
        // We must calculate the inverse matrix first
        try {
            int[][] invKey = getInverseMatrix(key);
            String decrypted = process(encrypted, invKey);
            System.out.println("Decrypted: " + decrypted);
        } catch (Exception e) {
            System.out.println("Error: " + e.getMessage());
        }
    }
}

File: python/Hill.py
import sys

def mod_inverse(a, m):
    """Finds modular multiplicative inverse of a under modulo m."""
    a = a % m
    for x in range(1, m):
        if (a * x) % m == 1:
            return x
    return None

def get_key_matrix_inverse(key_matrix):
    """Calculates the inverse of a 2x2 matrix modulo 26."""
    # Determinant = ad - bc
    det = key_matrix[0][0] * key_matrix[1][1] - key_matrix[0][1] * key_matrix[1][0]
    det = det % 26
    
    det_inv = mod_inverse(det, 26)
    if det_inv is None:
        raise ValueError("Key matrix is not invertible modulo 26.")
        
    # Inverse Matrix formula for 2x2:
    # [ d  -b ]
    # [ -c  a ]  * det_inv
    
    # Swap a and d, negate b and c
    inv_matrix = [
        [(key_matrix[1][1] * det_inv) % 26, (-key_matrix[0][1] * det_inv) % 26],
        [(-key_matrix[1][0] * det_inv) % 26, (key_matrix[0][0] * det_inv) % 26]
    ]
    return inv_matrix

def hill_cipher(text, key_matrix, mode='encrypt'):
    text = text.replace(" ", "").upper()
    # Padding if length is odd
    if len(text) % 2 != 0:
        text += "X"
        
    n = 2 # Size of the matrix (2x2)
    result = ""
    
    # Determine which matrix to use
    if mode == 'encrypt':
        matrix = key_matrix
    else:
        matrix = get_key_matrix_inverse(key_matrix)
        
    # Process text in blocks of 2
    for i in range(0, len(text), n):
        # Convert char block to vector [x, y]
        vector = [ord(text[i]) - 65, ord(text[i+1]) - 65]
        
        # Matrix Multiplication: (Matrix * Vector) % 26
        # C1 = (K00*P0 + K01*P1) % 26
        # C2 = (K10*P0 + K11*P1) % 26
        enc_vector = [
            (matrix[0][0] * vector[0] + matrix[0][1] * vector[1]) % 26,
            (matrix[1][0] * vector[0] + matrix[1][1] * vector[1]) % 26
        ]
        
        result += chr(enc_vector[0] + 65) + chr(enc_vector[1] + 65)
        
    return result

# --- Main Execution ---
if __name__ == "__main__":
    # Key matrix [[3, 3], [2, 5]]
    # Determinant = 15 - 6 = 9 (Coprime to 26, so it works)
    key = [[3, 3], [2, 5]]
    message = "HELP"
    
    try:
        encrypted = hill_cipher(message, key, mode='encrypt')
        decrypted = hill_cipher(encrypted, key, mode='decrypt')
        
        print(f"Original:  {message}")
        print(f"Encrypted: {encrypted}")
        print(f"Decrypted: {decrypted}")
    except ValueError as e:
        print(e)